agent/structs/structs.go

```diff
@@ -2391,3 +2410,15 @@ func (r *KeyringResponses) Add(v interface{}) {
 func (r *KeyringResponses) New() interface{} {
 	return new(KeyringResponses)
 }
+
+type UpstreamDownstream struct {
+	Upstream   ServiceName
+	Downstream ServiceName
+
+	// Upstream/Downstream pairs come from individual service registrations, and they can be updated individually.
+	// Refs stores the registrations that contain this pairing.
+	// When there are no remaining Refs, the UpstreamDownstream can be deleted.
+	Refs map[string]bool
```

Comment:
The bool is never used here to represent two different states right?

If not the canonical way to represent a set in Go is to use `map[type]struct{}` since an anonymous struct takes zero bytes of memory and doesn't need to be set to some arbitrary value like `true` to indicate presence.

Reply:
Is it also worth pointing out explicitly that as well all MemDB resident structs the map should be treated as immutable after inserting or reading from DB and only modified by deep copying the entire `UpstreamDownstream`. I realise you did this here in the PR but just a thought?